File: app/src/main/java/org/secuso/privacyfriendlytodolist/model/impl/TodoTaskImpl.java
/*
 This file is part of Privacy Friendly To-Do List.

 Privacy Friendly To-Do List is free software:
 you can redistribute it and/or modify it under the terms of the
 GNU General Public License as published by the Free Software Foundation,
 either version 3 of the License, or any later version.

 Privacy Friendly To-Do List is distributed in the hope
 that it will be useful, but WITHOUT ANY WARRANTY; without even
 the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 See the GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with Privacy Friendly To-Do List. If not, see <http://www.gnu.org/licenses/>.
 */

package org.secuso.privacyfriendlytodolist.model.impl;

import android.os.Parcel;
import android.util.Log;

import org.secuso.privacyfriendlytodolist.model.TodoList;
import org.secuso.privacyfriendlytodolist.model.TodoSubtask;
import org.secuso.privacyfriendlytodolist.model.TodoTask;
import org.secuso.privacyfriendlytodolist.model.database.entities.TodoTaskData;
import org.secuso.privacyfriendlytodolist.util.Helper;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

/**
 * Created by Sebastian Lutz on 12.03.2018.
 * <p>
 * Class to set up To-Do Tasks and its parameters.
 */

public class TodoTaskImpl extends BaseTodoImpl implements TodoTask {

    private static final String TAG = TodoTaskImpl.class.getSimpleName();


    /** Container for data that gets stored in the database. */
    private final TodoTaskData data;
    /** Important for the reminder service. */
    private boolean reminderTimeChanged = false;
    private boolean reminderTimeWasInitialized = false;
    private List<TodoSubtask> subtasks = new ArrayList<>();

    public TodoTaskImpl() {
        data = new TodoTaskData();
    }

    public TodoTaskImpl(TodoTaskData data) {
        this.data = data;
    }

    public TodoTaskImpl(Parcel parcel) {
        data = new TodoTaskData();
        data.setId(parcel.readInt());
        if (0 != parcel.readByte()) {
            data.setListId(parcel.readInt());
        } else {
            parcel.readInt();
            data.setListId(TodoList.DUMMY_LIST_ID);
        }
        data.setName(Objects.requireNonNullElse(parcel.readString(), ""));
        data.setDescription(Objects.requireNonNullElse(parcel.readString(), ""));
        data.setDone(parcel.readByte() != 0);
        data.setInRecycleBin(parcel.readByte() != 0);
        data.setProgress(parcel.readInt());
        data.setDeadline(parcel.readLong());
        data.setReminderTime(parcel.readLong());
        data.setListPosition(parcel.readInt());
        Priority priority = Priority.fromOrdinal(parcel.readInt());
        data.setPriority(Objects.requireNonNullElse(priority, Priority.DEFAULT_VALUE));
        parcel.readList(subtasks, TodoSubtaskImpl.class.getClassLoader());
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(data.getId());
        if (null != data.getListId()) {
            dest.writeByte((byte)1);
            dest.writeInt(data.getListId());
        } else {
            dest.writeByte((byte)0);
            dest.writeInt(0);
        }
        dest.writeString(data.getName());
        dest.writeString(data.getDescription());
        dest.writeByte((byte)(data.isDone() ? 1 : 0));
        dest.writeByte((byte)(data.isInRecycleBin() ? 1 : 0));
        dest.writeInt(data.getProgress());
        dest.writeLong(data.getDeadline());
        dest.writeLong(data.getReminderTime());
        dest.writeInt(data.getListPosition());
        dest.writeInt(data.getPriority().ordinal());
        dest.writeList(subtasks);
    }

    @Override
    public void setId(int id) {
        data.setId(id);
    }

    @Override
    public int getId() {
        return data.getId();
    }

    @Override
    public void setName(String name) {
        data.setName(name);
    }

    @Override
    public String getName() {
        return data.getName();
    }

    @Override
    public void setDescription(String description) {
        data.setDescription(description);
    }

    @Override
    public String getDescription() {
        return data.getDescription();
    }

    @Override
    public void setListId(Integer listId) {
        data.setListId(listId);
    }

    @Override
    public Integer getListId() {
        return data.getListId();
    }

    @Override
    public void setDeadline(long deadline) {
        data.setDeadline(deadline);
    }

    @Override
    public long getDeadline() {
        return data.getDeadline();
    }

    @Override
    public boolean hasDeadline() {
        return data.getDeadline() > 0;
    }

    @Override
    public void setListPosition(int position) {
        data.setListPosition(position);
    }

    @Override
    public int getListPosition() {
        return data.getListPosition();
    }

    @Override
    public void setSubtasks(List<TodoSubtask> subtasks) {
        this.subtasks = subtasks;
    }

    @Override
    public List<TodoSubtask> getSubtasks() {
        return subtasks;
    }

    @Override
    // This method expects the deadline to be greater than the reminder time.
    public DeadlineColors getDeadlineColor(long defaultReminderTime) {

        // The default reminder time is a relative value in seconds (e.g. 86400s == 1 day)
        // The user specified reminder time is an absolute timestamp

        DeadlineColors dDeadlineColor = DeadlineColors.BLUE;
        final long deadline = data.getDeadline();
        final long reminderTime = data.getReminderTime();
        if (!data.isDone() && deadline > 0) {

            long currentTimeStamp = Helper.getCurrentTimestamp();
            long remTimeToCalc = reminderTime > 0 ? deadline-reminderTime : defaultReminderTime;

            if ((currentTimeStamp >= (deadline - remTimeToCalc)) && (deadline > currentTimeStamp)) {
                dDeadlineColor = DeadlineColors.ORANGE;
            }
            else if (currentTimeStamp > deadline) {
                dDeadlineColor = DeadlineColors.RED;
            }
        }

        return dDeadlineColor;
    }

    @Override
    public void setPriority(Priority priority) {
        data.setPriority(priority);
    }

    @Override
    public Priority getPriority() {
        return data.getPriority();
    }

    @Override
    public void setProgress(int progress) {
        data.setProgress(progress);
    }

    @Override
    public int getProgress(boolean computeProgress) {
        if (computeProgress) {
            int progress;
            if (0 == subtasks.size()) {
                progress = isDone() ? 100 : 0;
            } else {
                int doneSubtasks = 0;
                for (TodoSubtask todoSubtask : subtasks) {
                    if (todoSubtask.isDone()) {
                        ++doneSubtasks;
                    }
                }
                progress = (doneSubtasks * 100) / subtasks.size();
            }
            setProgress(progress);
        }

        return data.getProgress();
    }

    public static final Creator<TodoTaskImpl> CREATOR =
        new Creator<>() {
            @Override
            public TodoTaskImpl createFromParcel(Parcel source) {
                return new TodoTaskImpl(source);
            }

            @Override
            public TodoTaskImpl[] newArray(int size) {
                return new TodoTaskImpl[size];
            }
        };

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void setReminderTime(long reminderTime) {
        long deadline = data.getDeadline();
        if(reminderTime > deadline && deadline > 0) {
            Log.i(TAG, "Reminder time must not be greater than the deadline.");
        }
        else {
            data.setReminderTime(reminderTime);
        }

        // check if reminder time was already set and now changed -> important for reminder service
        if (reminderTimeWasInitialized) {
            reminderTimeChanged = true;
        }
        reminderTimeWasInitialized = true;
    }

    @Override
    public long getReminderTime() {
        return data.getReminderTime();
    }

    @Override
    public boolean reminderTimeChanged() {
        return reminderTimeChanged;
    }

    @Override
    public void resetReminderTimeChangedStatus() {
        reminderTimeChanged = false;
    }

    @Override
    public void setAllSubtasksDone(boolean isDone) {
        for(TodoSubtask subtask : subtasks) {
            subtask.setDone(isDone);
        }
    }

    @Override
    public void setDone(boolean isDone) {
        data.setDone(isDone);
    }

    @Override
    public boolean isDone() {
        return data.isDone();
    }

    // A task is done if the user manually sets it done or when all subtasks are done.
    // If a subtask is selected "done", the entire task might be "done" if by now all subtasks are done.
    @Override
    public void doneStatusChanged() {
        boolean allSubtasksAreDone = true;
        for (TodoSubtask subtask : subtasks)
        {
            if (!subtask.isDone())
            {
                allSubtasksAreDone = false;
                break;
            }
        }

        if (allSubtasksAreDone != data.isDone()) {
            dbState = ObjectStates.UPDATE_DB;
        }

        data.setDone(allSubtasksAreDone);
    }

    @Override
    public void setInRecycleBin(boolean isInRecycleBin) {
        data.setInRecycleBin(isInRecycleBin);
    }

    @Override
    public boolean isInRecycleBin() {
        return data.isInRecycleBin();
    }

    @Override
    public boolean checkQueryMatch(String query, boolean recursive) {
        // no query? always match!
        if (query == null || query.isEmpty()) {
            return true;
        }
        String queryLowerCase = query.toLowerCase();
        if (data.getName().toLowerCase().contains(queryLowerCase)) {
            return true;
        }
        if (data.getDescription().toLowerCase().contains(queryLowerCase)) {
            return true;
        }
        if (recursive) {
            for (TodoSubtask subtask : subtasks) {
                if (subtask.checkQueryMatch(queryLowerCase)) {
                    return true;
                }
            }
        }
        return false;
    }

    @Override
    public boolean checkQueryMatch(String query) {
        return checkQueryMatch(query, true);
    }

    TodoTaskData getData() {
        return data;
    }
}

File: app/src/main/java/org/secuso/privacyfriendlytodolist/model/impl/TodoTaskImpl.kt
/*
Privacy Friendly To-Do List
Copyright (C) 2018-2025  Sebastian Lutz

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/
package org.secuso.privacyfriendlytodolist.model.impl

import android.os.Parcel
import android.os.Parcelable.Creator
import org.secuso.privacyfriendlytodolist.model.TodoSubtask
import org.secuso.privacyfriendlytodolist.model.TodoTask
import org.secuso.privacyfriendlytodolist.model.TodoTask.Priority
import org.secuso.privacyfriendlytodolist.model.TodoTask.RecurrencePattern
import org.secuso.privacyfriendlytodolist.model.TodoTask.ReminderState
import org.secuso.privacyfriendlytodolist.model.Urgency
import org.secuso.privacyfriendlytodolist.model.database.entities.TodoTaskData
import org.secuso.privacyfriendlytodolist.util.Helper
import java.util.Locale
import java.util.concurrent.TimeUnit

/**
 * Created by Sebastian Lutz on 12.03.2018.
 *
 * Class to set up To-Do Tasks and its parameters.
 */
class TodoTaskImpl : BaseTodoImpl, TodoTask {
    /** Container for data that gets stored in the database.  */
    val data: TodoTaskData

    private var subtasks: MutableList<TodoSubtask> = ArrayList()

    constructor() {
        data = TodoTaskData()
        // New item needs to be stored in database.
        requiredDBAction = RequiredDBAction.INSERT
    }

    constructor(data: TodoTaskData) {
        this.data = data
    }

    constructor(parcel: Parcel) {
        data = TodoTaskData()
        data.id = parcel.readInt()
        data.listId = parcel.readValue(Int::class.java.classLoader) as Int?
        data.name = parcel.readString()!!
        data.description = parcel.readString()!!
        data.creationTime = parcel.readLong()
        data.doneTime = parcel.readValue(Long::class.java.classLoader) as Long?
        data.isInRecycleBin = parcel.readByte() != 0.toByte()
        data.progress = parcel.readInt()
        data.deadline = parcel.readValue(Long::class.java.classLoader) as Long?
        data.recurrencePattern = RecurrencePattern.fromOrdinal(parcel.readInt())!!
        data.recurrenceInterval = parcel.readInt()
        data.reminderTime = parcel.readValue(Long::class.java.classLoader) as Long?
        data.reminderState = ReminderState.fromOrdinal(parcel.readInt())!!
        data.sortOrder = parcel.readInt()
        data.priority = Priority.fromOrdinal(parcel.readInt())!!
        parcel.readTypedList(subtasks, TodoSubtaskImpl.CREATOR)
        // The duplicated object shall not duplicate the RequiredDBAction. The original object shall
        // ensure that DB action gets done. So keep initial value RequiredDBAction.NONE.
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeInt(data.id)
        dest.writeValue(data.listId)
        dest.writeString(data.name)
        dest.writeString(data.description)
        dest.writeLong(data.creationTime)
        dest.writeValue(data.doneTime)
        dest.writeByte((if (data.isInRecycleBin) 1 else 0).toByte())
        dest.writeInt(data.progress)
        dest.writeValue(data.deadline)
        dest.writeInt(data.recurrencePattern.ordinal)
        dest.writeInt(data.recurrenceInterval)
        dest.writeValue(data.reminderTime)
        dest.writeInt(data.reminderState.ordinal)
        dest.writeInt(data.sortOrder)
        dest.writeInt(data.priority.ordinal)
        dest.writeTypedList(subtasks)
    }

    override fun setId(id: Int) {
        data.id = id
    }

    override fun getId(): Int {
        return data.id
    }

    override fun setCreationTime(creationTime: Long) {
        data.creationTime = creationTime
    }

    override fun getCreationTime(): Long {
        return data.creationTime
    }

    override fun setName(name: String) {
        data.name = name
    }

    override fun getName(): String {
        return data.name
    }

    override fun setDescription(description: String) {
        data.description = description
    }

    override fun getDescription(): String {
        return data.description
    }

    override fun setListId(listId: Int?) {
        data.listId = listId
    }

    override fun getListId(): Int? {
        return data.listId
    }

    override fun setDeadline(deadline: Long?) {
        data.deadline = deadline
    }

    override fun getDeadline(): Long? {
        return data.deadline
    }

    override fun hasDeadline(): Boolean {
        return data.deadline != null
    }

    override fun setRecurrencePattern(recurrencePattern: RecurrencePattern) {
        data.recurrencePattern = recurrencePattern
    }

    override fun getRecurrencePattern(): RecurrencePattern {
        return data.recurrencePattern
    }

    override fun setRecurrenceInterval(recurrenceInterval: Int) {
        data.recurrenceInterval = recurrenceInterval
    }

    override fun getRecurrenceInterval(): Int {
        return data.recurrenceInterval
    }

    override fun isRecurring(): Boolean {
        return data.recurrencePattern != RecurrencePattern.NONE
    }

    override fun setSortOrder(sortOrder: Int) {
        data.sortOrder = sortOrder
    }

    override fun getSortOrder(): Int {
        return data.sortOrder
    }

    override fun setSubtasks(subtasks: MutableList<TodoSubtask>) {
        this.subtasks = subtasks
    }

    override fun getSubtasks(): MutableList<TodoSubtask> {
        return subtasks
    }

    override fun getUrgency(reminderTimeSpan: Long): Urgency {
        var level = Urgency.Level.NONE
        var deadline = data.deadline
        var daysUntilDeadline: Long? = null
        if (!isDone() && deadline != null) {
            val now = Helper.getCurrentTimestamp()
            if (isRecurring()) {
                deadline = Helper.getNextRecurringDate(deadline,
                    data.recurrencePattern, data.recurrenceInterval, now)
            }
            val nowDay = TimeUnit.SECONDS.toDays(now)
            val deadlineDay = TimeUnit.SECONDS.toDays(deadline)
            daysUntilDeadline = deadlineDay - nowDay
            if (nowDay > deadlineDay) {
                level = Urgency.Level.EXCEEDED
            }
            else if (nowDay == deadlineDay) {
                level = Urgency.Level.DUE
            }
            else {
                // Set time-part to 00:00:00 to ensure that comparison with reminder time span
                // doesn't overlap with deadline-day. For example if deadline has time-part of
                // 12:00:00 and reminder time span is 0.5 day it would not get 'imminent' before the
                // deadline-day begins. But it should get 'imminent' 0.5 day before deadline day begins.
                deadline = Helper.changeTimePart(deadline)
                val reminderTime = data.reminderTime
                var finalReminderTimeSpan = reminderTimeSpan
                if (reminderTime != null && reminderTime < deadline) {
                    finalReminderTimeSpan = deadline - reminderTime
                }
                if ((deadline - finalReminderTimeSpan) <= now) {
                    level = Urgency.Level.IMMINENT
                }
            }
        }
        return Urgency(level, daysUntilDeadline)
    }

    override fun setPriority(priority: Priority) {
        data.priority = priority
    }

    override fun getPriority(): Priority {
        return data.priority
    }

    override fun setProgress(progress: Int) {
        data.progress = progress
    }

    override fun getProgress(): Int {
        return data.progress
    }

    override fun computeProgress(): Boolean {
        val computedProgress: Int
        @Suppress("LiftReturnOrAssignment")
        if (0 == subtasks.size) {
            computedProgress = if (isDone()) 100 else 0
        } else {
            var doneSubtasks = 0
            for (todoSubtask in subtasks) {
                if (todoSubtask.isDone()) {
                    ++doneSubtasks
                }
            }
            computedProgress = doneSubtasks * 100 / subtasks.size
        }
        val progressChanged = getProgress() != computedProgress
        if (progressChanged) {
            setProgress(computedProgress)
        }
        return progressChanged
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun setReminderTime(reminderTime: Long?) {
        data.reminderTime = reminderTime
        data.reminderState = ReminderState.INITIAL
    }

    override fun getReminderTime(): Long? {
        return data.reminderTime
    }

    override fun hasReminderTime(): Boolean {
        return data.reminderTime != null
    }

    override fun setReminderState(reminderState: ReminderState) {
        data.reminderState = reminderState
    }

    override fun getReminderState(): ReminderState {
        return data.reminderState
    }

    override fun setAllSubtasksDone(isDone: Boolean) {
        for (subtask in subtasks) {
            subtask.setDone(isDone)
        }
    }

    override fun setDone(isDone: Boolean) {
        data.doneTime = if (isDone) Helper.getCurrentTimestamp() else null
    }

    override fun isDone(): Boolean {
        return data.doneTime != null
    }

    override fun setDoneTime(doneTime: Long?) {
        data.doneTime = doneTime
    }

    override fun getDoneTime(): Long? {
        return data.doneTime
    }

    override fun updateDoneStatus(): Boolean {
        var doneStatusChanged = false
        if (subtasks.isNotEmpty()) {
            var allSubtasksAreDone = true
            for (subtask in subtasks) {
                if (!subtask.isDone()) {
                    allSubtasksAreDone = false
                    break
                }
            }
            doneStatusChanged = isDone() != allSubtasksAreDone
            if (doneStatusChanged) {
                setDone(allSubtasksAreDone)
                requiredDBAction = RequiredDBAction.UPDATE
            }
        }
        return doneStatusChanged
    }

    override fun setInRecycleBin(isInRecycleBin: Boolean) {
        data.isInRecycleBin = isInRecycleBin
    }

    override fun isInRecycleBin(): Boolean {
        return data.isInRecycleBin
    }

    override fun checkQueryMatch(query: String?, recursive: Boolean): Boolean {
        // no query? always match!
        if (query.isNullOrEmpty()) {
            return true
        }
        val queryLowerCase = query.lowercase(Locale.getDefault())
        if (data.name.lowercase(Locale.getDefault()).contains(queryLowerCase)) {
            return true
        }
        if (data.description.lowercase(Locale.getDefault()).contains(queryLowerCase)) {
            return true
        }
        if (recursive) {
            for (subtask in subtasks) {
                if (subtask.checkQueryMatch(queryLowerCase)) {
                    return true
                }
            }
        }
        return false
    }

    override fun checkQueryMatch(query: String?): Boolean {
        return checkQueryMatch(query, true)
    }

    override fun toString(): String {
        return "TodoTask(name=${getName()}, id=${getId()}, rp=${getRecurrencePattern()}, ri=${getRecurrenceInterval()})"
    }

    companion object {
        @JvmField
        val CREATOR = object : Creator<TodoTask> {
            override fun createFromParcel(parcel: Parcel): TodoTask {
                return TodoTaskImpl(parcel)
            }

            override fun newArray(size: Int): Array<TodoTask?> {
                return arrayOfNulls(size)
            }
        }
    }
}
